File: c/tree/binary-tree-maximum-path-sum.c
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     struct TreeNode *left;
 *     struct TreeNode *right;
 * };
 */
int get_max(struct TreeNode *root, int *node)
{
    if(root == NULL)
        return 0;

    int left = get_max(root->left,node);
    int right = get_max(root->right,node);
    int ret;
    int tmp;

    ret = root->val;
    ret += left > right ? left : right;

    tmp = root->val + left + right;
    *node = tmp > *node ? tmp : *node;

    return ret < 0 ? 0 : ret;
}

int maxPathSum(struct TreeNode* root) {
    int node = INT_MIN;

    get_max(root,&node);

    return node;
}

File: python/tree/binary-tree-maximum-path-sum.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def max_sum(self,root):
        if not root:
            return (0,-sys.maxint)

        left = self.max_sum(root.left)
        right = self.max_sum(root.right)

        retval = max(left[0]+root.val, right[0]+root.val,0)
        maxval = max(left[1],right[1],root.val+left[0]+right[0])

        return (retval,maxval)

    def maxPathSum(self, root):
        _, nodeval = self.max_sum(root)
        return nodeval
